mail: allow finegrained control of mail domains

mkg20001 opened the pull request:
This changes several things
    
    - mailserver.domains is an attrset with a submodule
    - mailserver.domains = ["example.tld"] is transparently mapped to
    mailserver.domains."example.tld" = { enable = true; }
    - deperate domains as list, make attrset the default in docs
    - the user can now disable (enabled by default) autoconfig
    mailserver.domains."example.tld".autoconfig = false;
    - update docs to reflect the changes
    
    Note: this is fully backwards compatible, but using a list is now discouraged
    and deperacted
    
    fixes PL-130003


@flyingcircusio/release-managers

## Release process

Impact:
- none

Changelog:
- mailserver.domains is now an attrset
- mailserver.domains."domain".autoconfig/primary can now be used to change whether autoconfig is enabled/which domain is primary
- map mailserver.domains (list) to mailserver.domains (attrset) to be backwards compatible
- add message about mailserver.domains (list) being deperacted

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  -  fix cert renewal caused by domain where autoconfig went unused, but was still enabled
- [x] Security requirements tested? (EVIDENCE)
  - tested on test vm

dpausp commented on nixos/roles/mailserver.nix:
```diff
@@ -91,6 +92,7 @@ in
         Mailout on all nodes in this RG/location.
       '';
 
+      # shorthand for directly creating domains in
```
comment not finished?

dpausp commented on nixos/roles/mailserver.nix:
```diff
@@ -99,9 +101,41 @@ in
+        });
+
+        default = (listToAttrs (map (domain: (nameValuePair domain { enable = true; })) roles.mailserver.domains))
+          // (if roles.mailserver.domains != [] then { ${elemAt roles.mailserver.domains 0} = { enable = true; primary = true; }; } else {});
```
use `optionalAttrs` here instead of if...else

dpausp replied:
head instead of elemAt ... 0

dpausp commented on nixos/roles/mailserver.nix:
```diff
@@ -99,9 +101,41 @@ in
 
+      # this allows finegrained control over each domain
+      # for example domain."test.fcio.net".autoconfig = false;
+      domain = mkOption {
```
can this be set in JSON config?

mkg20001 replied:
```
{"domain":{"domain.tld":{"autoconfig":false}}}
```

Does it automatically take all options from JSON or anything specific necesarry?

mkg20001 replied:
Tested, seems so it does it automagically

dpausp commented on nixos/services/mail/default.nix:
```diff
@@ -16,7 +17,12 @@ let
   primaryDomain =
-    if role.domains != [] then elemAt role.domains 0 else "example.org";
+    let firstPrimary =
+      elemAt (attrNames (filterAttrs (domain: config: config.enable && config.primary) role.domain)) 0;
```
head instead of elemAt ... 0

dpausp reviewed (changes requested):
We also need some documentation for the new options in doc/src/mailserver.rst.

mkg20001 committed:
mail: allow finegrained control of mail domains

This changes several things

- mailserver.domains is an attrset with a submodule
- mailserver.domains = ["example.tld"] is transparently mapped to
mailserver.domains."example.tld" = { enable = true; }
- deperate domains as list, make attrset the default in docs
- the user can now disable (enabled by default) autoconfig
mailserver.domains."example.tld".autoconfig = false;
- update docs to reflect the changes

Note: this is fully backwards compatible, but using a list is now discouraged
and deperacted

fixes PL-130003

dpausp commented:
@mkg20001 conflicts with your previous PR, please rebase
